### ISSUE TITLE
Download latest build here

### ISSUE THREAD

**Issue by MolotovCherry** [MolotovCherry/virtual-display-rs#115]
The latest master build has a lot of features and fixes, and is stable.

Between the current stable release and latest master, these are the following feature changes:
- no admin access required for updates to monitor
- working cli tool (thanks @kylewlacy!)
- optimized performance under cpu load
- lots of miscellaneous improvements and fixes
- hardware cursor support (for remote desktop apps, this is a big plus!)
- up to 16 monitors (minus your only existing monitors, usable number might be slightly lower)
- complete python bindings (python 3.12 only!)
  - examples on how to use it are [here](https://github.com/MolotovCherry/virtual-display-rs/blob/master/examples/monitor_control.py)
- windows service which does the job of starting persisted monitors on boot (disable or don't install if you're not interested in persistence)
  - place `vdd-user-session-service.exe` wherever you want
  - run as admin `.\vdd-user-session-service.exe --install` to install service
  - to uninstall, run as admin `.\vdd-user-session-service.exe --uninstall`

Note: there is currently no gui in master release

[Download here](https://github.com/MolotovCherry/virtual-display-rs/actions/runs/8620193705) - **updated 4/9/24**

To install:
- uninstall the driver completely
- Follow the manual driver installation steps [here](https://github.com/MolotovCherry/virtual-display-rs?tab=readme-ov-file#how-to-install-portable-version) (you can skip the registry file and certificate steps)

**Comment by psychrabi** [comment]
![image](https://github.com/MolotovCherry/virtual-display-rs/assets/17595403/7e8ad3af-4a50-443c-a627-6ed9c4204b44)
Didnot see the vdd-user-session-service.exe in the folder

**Comment by MolotovCherry** [comment]
> Didnot see the vdd-user-session-service.exe in the folder
How did you download it? I see it in the zip from the download link provided
![image](https://github.com/MolotovCherry/virtual-display-rs/assets/13651622/7341d3e6-94d9-449b-9515-0c7a60317799)

**Comment by psychrabi** [comment]
i downloaded from here
![image](https://github.com/MolotovCherry/virtual-display-rs/assets/17595403/9b963f3d-b9f2-43c0-b5db-d6249f24ada9)

**Comment by psychrabi** [comment]
I redownloaded the driver now and i see the service.exe now
![image](https://github.com/MolotovCherry/virtual-display-rs/assets/17595403/8563c097-1e4e-4b49-96d6-2755f7fc3e07)

**Comment by fabiomilson** [comment]
The download button on the job is not showing, why? I can't download the driver :(

**Comment by MolotovCherry** [comment]
> The download button on the job is not showing, why? I can't download the driver :(

You can click the name of the artifact to initiate download

![image](https://github.com/user-attachments/assets/a83eaeb3-5208-4f59-8bfb-84b7da302bb8)

**Comment by yaoleiliu** [comment]
I successfully installed it, but only the display adapter was found, not the monitor.

**Comment by tristan-k** [comment]
Where is the documentation on how to use `vdd-user-session-service.exe`? The help flag isnt really telling much either:

```
 .\vdd-user-session-service.exe --help
Usage: vdd-user-session-service.exe [OPTIONS]

Options:
  -i, --install    Install the service
  -u, --uninstall  Uninstall the service
  -h, --help       Print help
  -V, --version    Print version
```

**Comment by MolotovCherry** [comment]
> * .\vdd-user-session-service.exe --install

@tristan-k As it says in the first post:
As admin, run `.\vdd-user-session-service.exe --install` to install it
To uninstall it run as admin `.\vdd-user-session-service.exe --uninstall`

**Comment by yaoleiliu** [comment]
./vdd-user-session-service.exe -i

**Comment by tristan-k** [comment]
So `vdd-user-session-service.exe` is just for installing the service but how do I create and edit virtual display?

**Comment by yaoleiliu** [comment]
use the cli.exe to add monitor

**Comment by MolotovCherry** [comment]
@yaoleiliu I'll look into HDR when I have time, but right now there's no ETA since I've been very busy.

@tristan-k Use `virtual-display-driver-cli.exe` to control the virtual displays. How to use it is displayed in the help commands, including subcommands.
```
.\virtual-display-driver-cli.exe --help
.\virtual-display-driver-cli.exe <subcommand> --help
```

**Comment by tristan-k** [comment]
Thanks. I just installed certificate with `./install-cert.bat` the service with `.\vdd-user-session-service.exe --install` and then tried adding a virtual display with `.\virtual-display-driver-cli.exe add 1920x1080` but I do get an error.

```
.\virtual-display-driver-cli.exe add 1920x1080
Error: Failed to connect to Virtual Display Driver; please ensure the driver is installed and working

Caused by:
   0: Failed to connect to driver: Failed to open pipe: Das System kann die angegebene Datei nicht finden. (os error 2)
   1: Failed to open pipe: Das System kann die angegebene Datei nicht finden. (os error 2)
   2: Das System kann die angegebene Datei nicht finden. (os error 2)
```

**Comment by yaoleiliu** [comment]
> Thanks. I just installed the certificate with `.\install-cert.bat`, the service with `.\vdd-user-session-service.exe --install` and then tried adding a virtual display with `.\virtual-display-driver-cli.exe add 1920x1080` but I do get an error.
> 
> ```
> .\virtual-display-driver-cli.exe add 1920x1080
> Error: Failed to connect to Virtual Display Driver; please ensure the driver is installed and working
> 
> Caused by:
>    0: Failed to connect to driver: Failed to open pipe: Das System kann die angegebene Datei nicht finden. (os error 2)
>    1: Failed to open pipe: Das System kann die angegebene Datei nicht finden. (os error 2)
>    2: Das System kann die angegebene Datei nicht finden. (os error 2)
> ```

you need install the driver after install-cert.bat

**Comment by tristan-k** [comment]
Okay, so the prodecure is:

```
.\nefconc.exe --remove-device-node --hardware-id Root\VirtualDisplayDriver --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318"
.\nefconc.exe --create-device-node --class-name Display --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318" --hardware-id Root\VirtualDisplayDriver
.\nefconc.exe --install-driver --inf-path ".\VirtualDisplayDriver.inf"
.\install-cert.bat
.\vdd-user-session-service.exe --install
.\virtual-display-driver-cli.exe add 1920x1080
```

**Comment by yaoleiliu** [comment]
> Okay, so the prodecure is:
> 
> ```
> .\nefconc.exe --remove-device-node --hardware-id Root\VirtualDisplayDriver --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318"
> .\nefconc.exe --create-device-node --class-name Display --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318" --hardware-id Root\VirtualDisplayDriver
> .\nefconc.exe --install-driver --inf-path ".\VirtualDisplayDriver.inf"
> .\install-cert.bat
> .\vdd-user-session-service.exe --install
> .\virtual-display-driver-cli.exe add 1920x108
> ```
> 

no
the prodecure is:
 .\install-cert.bat 
.\nefconc.exe --remove-device-node --hardware-id Root\VirtualDisplayDriver --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318" .\nefconc.exe --create-device-node --class-name Display --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318" --hardware-id Root\VirtualDisplayDriver .\nefconc.exe --install-driver --inf-path ".\VirtualDisplayDriver.inf".\vdd-user-session-service.exe --install .\virtual-display-driver-cli.exe add 1920x108

**Comment by MolotovCherry** [comment]
> Okay, so the prodecure is:
> 
> ```
> .\nefconc.exe --remove-device-node --hardware-id Root\VirtualDisplayDriver --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318"
> .\nefconc.exe --create-device-node --class-name Display --class-guid "4D36E968-E325-11CE-BFC1-08002BE10318" --hardware-id Root\VirtualDisplayDriver
> .\nefconc.exe --install-driver --inf-path ".\VirtualDisplayDriver.inf"
> .\install-cert.bat
> .\vdd-user-session-service.exe --install
> .\virtual-display-driver-cli.exe add 1920x108
> ```
> 

If you installed the driver before installing the certificate, it will cause a system error. To fix that, you need to completely uninstall the driver again, make sure the certificate is installed, then install the driver again.

**Comment by tristan-k** [comment]
Got it. Now it kinda works. Though I'm still unable to connect to Sunshine because I'm using [gpu partitioning with hyper-v](https://github.com/jamesstringerparsec/Easy-GPU-PV) and ran into an issue .[ According to the sunshine devs](https://github.com/LizardByte/Sunshine/issues/1937#issuecomment-1872558099) the display has to be connected to the GPU. But this is different issue proably not related to this project.

**Comment by tristan-k** [comment]
Followup question: Is there an easy way with `nefconc.exe` to uninstall the certificates previously installed?

**Comment by MolotovCherry** [comment]
> Followup question: Is there an easy way with `nefconc.exe` to uninstall the certificates previously installed?

@tristan-k No, nefcon is only for drivers. You have to go into "Manage User Certificates" or "Manage Computer Certificates" and manually delete them.

**Comment by yaoleiliu** [comment]
Is this latest driver only for win11?I always get 31 error when installing on win10

**Comment by God-damnit-all** [comment]
So, just to be clear, the latest GUI is not compatible at all with the latest driver?

**Comment by MolotovCherry** [comment]
> So, just to be clear, the latest GUI is not compatible at all with the latest driver?

That's correct. It isn't compatible with any GUI currently. Also, any future GUI will be completely new and not based on the old one.

**Comment by yeongjet** [comment]
What modes does it support?  There is no informations in `virtual-display-driver-cli` help command.
And does it support custom resolution that unusual? like `7890x3210`

**Comment by MolotovCherry** [comment]
@yeongjet Use `virtual-display-driver-cli help <subcommand>`

Yes, you can type in any resolution you want

**Comment by yeongjet** [comment]
> [@yeongjet](https://github.com/yeongjet) Use `virtual-display-driver-cli help <subcommand>`
> 
> Yes, you can type in any resolution you want

Thank you!
